app/pulp/app/viewsets/task.py

```diff
@@ -0,0 +1,46 @@
+from django_filters.rest_framework import filterset
+
+from pulp.app.models import Task, Worker
+from pulp.app.serializers import TaskSerializer, WorkerSerializer
+from pulp.app.viewsets import NamedModelViewSet
+from pulp.app.viewsets.custom_filters import CharInFilter
+
+from rest_framework.decorators import detail_route
+from rest_framework.response import Response
+from rest_framework import status
+
+
+class TaskFilter(filterset.FilterSet):
+    tags = CharInFilter(name='tag__name', lookup_expr='in')
+
+    class Meta:
+        model = Task
+        fields = ['state', 'worker__name', 'tags']
+
+
+class TaskViewSet(NamedModelViewSet):
+    queryset = Task.objects.all()
+    serializer_class = TaskSerializer
+    endpoint_name = 'tasks'
+    filter_class = TaskFilter
+    http_method_names = ['get', 'options']
+
+    @detail_route(methods=['post'])
+    def cancel(self, request, pk=None):
+        task = self.get_object()
+        if task.state == Task.RUNNING or task.state == Task.WAITING:
+            # TODO call a controller method to cancel task
```

Comment:
Thinking about this a little more, I wonder if this would be better to have on the model itself, even if it's a temporary home while we think about moving it to a controller.

Most of the logic here is about managing the state of the task itself, and whether a particular change is ok. That clearly is the domain of the model.

The only part of cancellation that I think might not be a good fit for the model is a call to celery to actively carry out cancellation. Calling that from the model wouldn't be the worst thing, and would probably be better than calling it from here (and everywhere else that wants to cancel a task).

What do you think?